### PR TITLE
Exunit simplified

### PR DESCRIPTION
**Pull request by iilyak**
## Overview

Eunit testing framework has a number of deficiencies especially in the cases when mocking is involved. This PR introduces Elixir based ExUnit into our testing stack. The PR doesn't include tests themselves. The tests would be included in separate PRs.

## Testing recommendations

Create `src/couch/test/exunit/crud_test.exs` with following content
```
defmodule Couch.Test.CRUD do
  use Couch.Test.ExUnit.Case
  alias Couch.Test.Utils

  alias Couch.Test.Setup

  alias Couch.Test.Setup.Step

  def with_db(context, setup) do
    setup =
      setup
      |> Setup.Common.httpd_with_db()
      |> Setup.run()

    context =
      Map.merge(context, %{
        db_name: setup |> Setup.get(:db) |> Step.Create.DB.name(),
        base_url: setup |> Setup.get(:start) |> Step.Start.clustered_url(),
        user: setup |> Setup.get(:admin) |> Step.User.name()
      })

    {context, setup}
  end

  describe "Database CRUD using Fabric API" do
    @describetag setup: &__MODULE__.with_db/2
    test "Create DB", ctx do
      IO.puts("base_url: #{ctx.base_url}")
      IO.puts("admin: #{ctx.user}")
      IO.puts("db_name: #{ctx.db_name}")
    end
  end
end
```
Create `src/couch/test/exunit/test_helper.exs` with the following content:
```
ExUnit.configure(formatters: [JUnitFormatter, ExUnit.CLIFormatter])
ExUnit.start()`
```

## Related Issues or Pull Requests

- old PR https://github.com/apache/couchdb/pull/2036 (the current PR is simplified by removing notion of testing adapter)
- RFC https://github.com/apache/couchdb-documentation/pull/415
- [Discussion on mailing list](https://lists.apache.org/thread.html/f842ca637f7cb06b34af699a793cab0a534e65970172e8117bf0b228@%3Cdev.couchdb.apache.org%3E)

## Checklist

- [x] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -173,6 +173,20 @@ eunit: couch
 	  $(REBAR) -r eunit $(EUNIT_OPTS) apps=$$dir || exit 1; \
 	done
 
+.PHONY: exunit
+# target: exunit - Run ExUnit tests
+exunit: export BUILDDIR = $(shell pwd)
+exunit: export MIX_ENV=test
+exunit: export ERL_LIBS = $(shell pwd)/src
+exunit: export ERL_AFLAGS = -config $(shell pwd)/rel/files/eunit.config
+exunit: export COUCHDB_QUERY_SERVER_JAVASCRIPT = $(shell pwd)/bin/couchjs $(shell pwd)/share/server/main.js
```
Comment (garrensmith) [inline Makefile]: would it be possible to add a new command where we can specify an app or file to test.

File: mix.exs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  # Run "mix help deps" to learn about dependencies.
+  defp deps() do
+    [
+      # {:dep_from_hexpm, "~> 0.3.0"},
+      {:httpotion, "~> 3.0", only: [:dev, :test], runtime: false},
+      {:jiffy, "~> 0.15.2", only: [:dev, :test], runtime: false},
+      {:credo, "~> 1.0.0", only: [:dev, :test], runtime: false},
+      {:junit_formatter, "~> 3.0", only: [:test], runtime: false}
+
+      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
```
Comment (garrensmith) [inline mix.exs]: We can probably remove this comment.

File: mix.exs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  # Specifies which paths to compile per environment.
+  defp elixirc_paths(:test), do: ["test/elixir/lib", "test/elixir/test/support"]
+  defp elixirc_paths(_), do: ["test/elixir/lib"]
+
+  # Run "mix help deps" to learn about dependencies.
+  defp deps() do
+    [
+      # {:dep_from_hexpm, "~> 0.3.0"},
```
Comment (garrensmith) [inline mix.exs]: We can probably remove this

File: test/elixir/README.md [hunk 4]
```diff
@@ -111,3 +111,146 @@ X means done, - means partially
...
+      Map.merge(context, %{
+        db_name: setup |> Setup.get(:db) |> Step.Create.DB.name(),
+        base_url: setup |> Setup.get(:start) |> Step.Start.clustered_url(),
+        user: setup |> Setup.get(:admin) |> Step.User.name()
+      })
+
+    {context, setup}
+  end
+
+  describe "Database CRUD using Fabric API" do
```
Comment (garrensmith) [inline test/elixir/README.md]: Really nice documentation, this helps a lot.

File: test/elixir/lib/setup.ex [hunk 5]
```diff
@@ -0,0 +1,86 @@
+defmodule Couch.Test.Setup do
+  @moduledoc """
+  Allows to chain setup functions.
+  Example of using:
+
+  ```
+      @tag setup: %Setup{}
+          |> Setup.Start.new([:chttpd])
+          |> Setup.Adapter.new(adapter)
```
Comment (garrensmith) [inline test/elixir/lib/setup.ex]: I think you have removed the adapter code.

File: Makefile [hunk 6]
```diff
@@ -173,6 +173,20 @@ eunit: couch
 	  $(REBAR) -r eunit $(EUNIT_OPTS) apps=$$dir || exit 1; \
 	done
 
+.PHONY: exunit
```
Comment (wohali) [inline Makefile]: Don't forget Makefile.win, please.

File: test/elixir/lib/setup.ex (hunk 5 above)
Comment (iilyak) [reply test/elixir/lib/setup.ex]: Good catch. This part of documentation was not updated.

File: Makefile (hunk 1 above)
Comment (iilyak) [reply Makefile]: Fixed. The syntax is:
```
make exunit tests=src/couch/test/exunit/crud_test.exs,src/fabric/test/exunit/crud_test.exs:line
```

Comment (garrensmith) [reply Makefile]: awesome. That is useful

**Comment by nickva** [comment]
Very nice work, @iilyak !

I noticed there was a formatting error in mix, let's fix that.  Was also wondering if there is any way to avoid renaming all the paths to the eunit tests. It would make things harder, that's fine, but didn't know what reason was. Maybe also squash some commits

**Comment by iilyak** [comment]
> @nickva if there is any way to avoid renaming all the paths to the eunit tests. If it would make things harder, that's fine, but didn't know what reason was. 

I moved eunit tests for better readability. I personally find it easier when tests are organized in separate directories. It is possible to keep eunit tests in the old location. In this case the structure would be:

```
<app>/test/some_test.erl
<app>/test/exunit/test_helper.exs
<app>/test/exunit/some_test.exs
```

Please let me know if you want me to change directory structure to keep old eunit where they were.

**Comment by iilyak** [comment]
> @nickva Maybe also squash some commits

I was trying to avoid rebase when review is in progress. I can do a cleanup.

File: mix.exs (hunk 2 above)
Comment (iilyak) [reply mix.exs]: fixed

File: mix.exs (hunk 3 above)
Comment (iilyak) [reply mix.exs]: fixed

File: test/elixir/lib/setup.ex (hunk 5 above)
Comment (iilyak) [reply test/elixir/lib/setup.ex]: fixed

File: Makefile (hunk 6 above)
Comment (iilyak) [reply Makefile]: Makefile.win is updated (unfortunately I don't have access to windows machine to test it)

**Comment by iilyak** [comment]
@nickva I did a rebase to reduce number of commits.

**Comment by nickva** [comment]
> > @nickva if there is any way to avoid renaming all the paths to the eunit tests. If it would make things harder, that's fine, but didn't know what reason was.
> 
> I moved eunit tests for better readability. I personally find it easier when tests are organized in separate directories. It is possible to keep eunit tests in the old location. In this case the structure would be:
> 
> ```
> <app>/test/some_test.erl
> <app>/test/exunit/test_helper.exs
> <app>/test/exunit/some_test.exs
> ```
> 
> Please let me know if you want me to change directory structure to keep old eunit where they were.

Ah ok, `eunit` and `exunit` folders do look better.

**Comment by nickva** [comment]
The commits look good. Thanks for rebasing.

There is some issue with mix still think:

```
All dependencies up to date
mix format --check-formatted
mix credo
[31m[1mUnchecked dependencies for environment dev:[0m
[31m[1m* jiffy (Hex package)[0m
[31m[1m  lock mismatch: the dependency is out of date. To fetch locked version run "mix deps.get"[0m
[31m[1m* httpotion (Hex package)[0m
[31m[1m  the dependency is not available, run "mix deps.get"[0m
[31m[1m* credo (Hex package)[0m
[31m[1m  the dependency is not available, run "mix deps.get"[0m
[31m[1m** (Mix) Can't continue due to errors on dependencies[0m
```

**Comment by iilyak** [comment]
@nickva I changed the definition for `jiffy` dependency. Could you try again?

**Comment by nickva** [comment]
@iilyak 

I gave it a try and still get the error:

```
==> asf-master (compile)
mix format --check-formatted
mix credo
Unchecked dependencies for environment test:
* jiffy (../jiffy)
  the dependency is not available
* httpotion (Hex package)
  the dependency is not available, run "mix deps.get"
* credo (Hex package)
  the dependency is not available, run "mix deps.get"
** (Mix) Can't continue due to errors on dependencies
```

Both locally and CI as well it seems

**Comment by nickva** [comment]
Almost there. I see a few errors like

```
[31m** (File.Error) could not read file "test/data/lorem.txt": no such file or directory[0m
``` 

In Travis logs. It doesn't seem like a timing issue. Maybe we changed some default paths or something of that sort.

**Comment by iilyak** [comment]
@nickva Tests pass on Jenkins. Let me know if you want me to rebase.

**Pull request merged**